sycl/source/detail/graph_impl.hpp

```diff
@@ -620,11 +645,21 @@ class graph_impl {
   /// Controls whether we skip the cycle checks in makeEdge, set by the presence
   /// of the no_cycle_check property on construction.
   bool MSkipCycleChecks = false;
+
+  /// Insert node into list of root nodes.
+  /// @param Root Node to add to list of root nodes.
+  void addRoot(const std::shared_ptr<node_impl> &Root);
 };
 
 /// Class representing the implementation of command_graph<executable>.
 class exec_graph_impl {
 public:
+  using ReadLock = std::shared_lock<std::shared_mutex>;
+  using WriteLock = std::unique_lock<std::shared_mutex>;
+
+  /// Protects all the fields that can be changed by class' methods.
+  mutable std::shared_mutex MMutex;
```

Comment:
Since each executable and modifiable graph has its own mutex, and there is a 1:n mapping of modifiable to executable graphs I see the potential for a deadlock. Because the order how we lock the graph becomes important. Do we have a test in place that would catch this?

Reply:
Yes your understanding is correct. It might help to document somewhere that we'll have to take the locks always in the same order, e.g. modifiable first then executable. This could help reducing the risk of introducing a deadlock in the future.

Reply:
I added a comment.